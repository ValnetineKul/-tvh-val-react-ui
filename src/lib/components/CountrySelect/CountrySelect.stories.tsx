import React, { useState } from 'react';
import CountrySelect from './index';
import type { CountryItem } from './CountrySelect';
import type { StoryTemplate } from '../../story-utils';
import Button from '../Buttons/Button';
import Banner from '../Banner';
import { createMeta } from '../../story-utils';

type Props = React.ComponentProps<typeof CountrySelect>;

type Args = Omit<Props, 'onChange' | 'value'>;

const countries: CountryItem[] = [
  { code: 'BE', name: 'Belgium' },
  { code: 'BY', name: 'Belarus' },
  { code: 'GE', name: 'Georgia' },
  { code: 'RU', name: 'Russia' },
  { code: 'CY', name: 'Cyprus' },
];

const Template = ((args: Args) => {
  const [value, setValue] = useState<CountryItem | null>({ code: 'BE', name: 'Belgium' });
  return <CountrySelect {...args} onChange={(_, v) => setValue(v as CountryItem)} value={value} />;
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  options: countries,
  label: 'Label',
  placeholder: 'Select a value',
  disabled: false,
  fullWidth: false,
  required: false,
  readOnly: false,
};

Primitive.storyName = 'Single';

const ErrorTemplate = ((args: Args) => {
  const [value, setValue] = useState<CountryItem | null>(null);

  return (
    <CountrySelect
      id="select-id"
      {...args}
      onChange={(_, v) => setValue(v as CountryItem)}
      value={value}
      NoOptionsText={
        <Banner
          status="error"
          message="Could not load list. Please refresh."
          button={<Button label="Refresh" onClick={() => 'TO_DO'} variant="tertiary" />}
          direction="vertical"
        />
      }
    />
  );
}) as StoryTemplate<Args>;

export const Error = ErrorTemplate.bind({});
Error.args = {
  ...Primitive.args,
  options: [],
};

export default createMeta({
  title: 'Components/Selects/CountrySelect',
  component: CountrySelect,
});
