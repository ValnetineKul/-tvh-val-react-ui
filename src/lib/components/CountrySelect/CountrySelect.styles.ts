import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles({ name: 'CountrySelect' })(() => ({
  isInTextArea: {
    margin: 0,
  },
}));

export default useStyles;
