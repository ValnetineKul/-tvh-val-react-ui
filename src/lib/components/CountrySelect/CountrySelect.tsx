import type { FC } from 'react';
import React from 'react';
import type { CountryCode } from '../../types/common';
import FlagIcon from '../PhoneField/FlagIcon/FlagIcon';
import type { SelectProps } from '../Select';
import Select from '../Select';
import useStyles from './CountrySelect.styles';

export interface CountryItem {
  code: CountryCode;
  name: string;
}

export type CountrySelectProps = Omit<
  SelectProps<CountryItem, false>,
  'getOptionLabel' | 'getOptionStartIcon' | 'multiple'
>;

const CountrySelect: FC<CountrySelectProps> = ({ disabled = false, ...restProps }) => {
  const { classes, cx } = useStyles();

  return (
    <Select<CountryItem, false>
      disabled={disabled}
      getOptionLabel={(option) => option.name}
      getOptionStartIcon={(option, params) => (
        <FlagIcon
          className={cx({ [classes.isInTextArea]: params?.isInTextArea })}
          countryCode={option.code}
          country={option.name}
          disabled={disabled}
        />
      )}
      isOptionEqualToValue={(o, v) => {
        return o.code === v.code;
      }}
      {...restProps}
    />
  );
};

export default CountrySelect;
