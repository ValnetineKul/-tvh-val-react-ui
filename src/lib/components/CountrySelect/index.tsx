import CountrySelect from './CountrySelect';
import type { CountryItem, CountrySelectProps } from './CountrySelect';

export type { CountryItem, CountrySelectProps };
export default CountrySelect;
