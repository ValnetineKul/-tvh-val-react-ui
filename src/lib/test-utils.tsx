import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
