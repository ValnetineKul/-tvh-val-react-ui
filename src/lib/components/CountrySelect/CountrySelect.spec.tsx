import React from 'react';
import { render } from '../../test-utils';
import CountrySelect from './CountrySelect';

describe('CountrySelect', () => {
  it('Should render correctly', () => {
    const { container } = render(
      <CountrySelect label="label" options={[{ code: 'BE', name: 'Belgium' }]} onChange={jest.fn()} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
