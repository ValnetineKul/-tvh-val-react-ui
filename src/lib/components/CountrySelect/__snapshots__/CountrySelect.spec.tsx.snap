// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelect Should render correctly 1`] = `
<div
  class="MuiFormControl-root tss-16-TextField-withStartAdornment-ref tss-16-TextField-withLabel-ref tss-1rqzzt7-TextField-root-Select-formControl-TextField-withStartAdornment-TextField-withLabel mui-1nrlq1o-MuiFormControl-root"
>
  <label
    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary tss-mactna-InputLabel-root-TextField-label-Select-label mui-xlr99k-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink="false"
    for="id-input-label"
  >
    <span
      class="tss-2-Typography-weightRegular-ref tss-1qugfhp-Typography-root-Typography-weightRegular-InputLabel-label-Typography-body400"
    >
      label
    </span>
  </label>
  <div
    class="tss-bdw984-Surface-root"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd tss-16-TextField-inputRoot-ref tss-1wa0j4i-TextField-inputRoot-TextField-adornedStart-TextField-adornedEnd mui-1cflty8-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="tss-1sv2whm-TextField-input MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd tss-1i9zx85-Select-input mui-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
        id="id-input-label"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium tss-1tivgqf-InputAdornment-root mui-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Toggle select"
          class="MuiButtonBase-root tss-1b9x7sz-ButtonBase-root-ButtonBase-focusOut-IconButton-root mui-10d1a0h-MuiButtonBase-root"
          tabindex="-1"
          type="button"
        >
          <div
            class="tss-w1sjks-Icon-root-Icon-sizeMd"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
              color="inherit"
              focusable="false"
              viewBox="0 0 24 24"
            >
              angleDown.svg
            </svg>
          </div>
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="tss-j7rywy-TextField-notchedOutline tss-16-TextField-notchedOutline-ref MuiOutlinedInput-notchedOutline mui-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="mui-nnbavb"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="tss-19z6szl-TextField-helpers"
  />
</div>
`;
